File: web-mobile/pagamento-semplice/pagamento-ricorrente/java/notifica.java
import java.util.Map;
import java.util.HashMap;
import java.security.MessageDigest;

public class notifica {

  public static void main(String[] args) throws Exception {

    // Chiave segreta
    String CHIAVESEGRETA = "TLGHTOWIZXQPTIZRALWKG"; // Sostituire con il valore fornito da CartaSi

    // Questi sono i parametri in ingresso della richiesta
    Map < String, String > paramFromRequest = new HashMap < String, String > ();
    paramFromRequest.put("codTrans", "");
    paramFromRequest.put("esito", "OK");
    paramFromRequest.put("importo", "");
    paramFromRequest.put("divisa", "");
    paramFromRequest.put("data", "");
    paramFromRequest.put("orario", "");
    paramFromRequest.put("codAut", "");
    paramFromRequest.put("mac", "58815356e9052f7d6a355a399d1d8edfc4a58bd7");  

    // Calcolo MAC
    String stringaMac = "codTrans=" + paramFromRequest.get("codTrans") +
            "esito=" + paramFromRequest.get("esito") +
            "importo=" + paramFromRequest.get("importo") +
            "divisa=" + paramFromRequest.get("divisa") +
            "data=" + paramFromRequest.get("data") +
            "orario=" + paramFromRequest.get("orario") +
            "codAut=" + paramFromRequest.get("codAut") +
            CHIAVESEGRETA;

    String mac = hashMac(stringaMac);

    // Verifico corrispondenza tra MAC calcolato e parametro mac in ingresso  
    if (!mac.equals(paramFromRequest.get("mac"))) {
      throw new Exception("Errore MAC: " + mac + " non corrisponde a " + paramFromRequest.get("mac"));
    }

    // Nel caso in cui non ci siano errori gestisco il parametro esito
    if ("OK".equals(paramFromRequest.get("esito"))) {
      System.out.println("OK, pagamento avvenuto, preso riscontro");
    } else {
      System.out.println("KO, pagamento non avvenuto, preso riscontro");
    }

  }  

  public static String hashMac(String stringaMac) throws Exception {
    MessageDigest digest = MessageDigest.getInstance("SHA-1");
    byte[] in = digest.digest(stringaMac.getBytes("UTF-8"));

    final StringBuilder builder = new StringBuilder();
    
    for(byte b : in) {
      builder.append(String.format("%02x", b));
    }

    return builder.toString();
  }

}

File: web-mobile/pagamento-semplice/pagamento-ricorrente/php/notifica.php
<?php

// Chiave segreta
$chiaveSegreta = "TLGHTOWIZXQPTIZRALWKG"; // Sostituire con il valore fornito da CartaSi

// Controllo che ci siano tutti i parametri di ritorno obbligatori per calcolare il MAC
$requiredParams = array('codTrans', 'esito', 'importo', 'divisa', 'data', 'orario', 'codAut', 'mac');
foreach ($requiredParams as $param) {
    if (!isset($_REQUEST[$param])) {
        echo 'Paramentro mancante ' . $field;
        header('500 Internal Server Error', true, 500);
        exit;
    }
}

// Calcolo MAC
$macCalculated = sha1('codTrans=' . $_REQUEST['codTrans'] .
        'esito=' . $_REQUEST['esito'] .
        'importo=' . $_REQUEST['importo'] .
        'divisa=' . $_REQUEST['divisa'] .
        'data=' . $_REQUEST['data'] .
        'orario=' . $_REQUEST['orario'] .
        'codAut=' . $_REQUEST['codAut'] .
        $chiaveSegreta
);

// Verifico corrispondenza tra MAC calcolato e parametro mac di ritorno
if ($macCalculated != $_REQUEST['mac']) {
    echo 'Errore MAC: ' . $macCalculated . ' non corrisponde a ' . $_REQUEST['mac'];
    header('500 Internal Server Error', true, 500);
    exit;
}

// Nel caso in cui non ci siano errori gestisco il parametro esito
if($_REQUEST['esito'] == 'OK'){
    header('OK, pagamento avvenuto, preso riscontro', true, 200);
} else {
    header('KO, pagamento non avvenuto, preso riscontro', true, 200);
}
